### ISSUE TITLE
Apparent deadlock during startup and shutdown using WinGUI

### ISSUE THREAD

**Issue by clangen** [Bill-Gray/PDCursesMod#212]
Hello!

I'm working on upgrading my project to use the most recent version of PDCursesMod, but have run into what appears to be a difficult-to-reproduce deadlock using WinGUI. This manifests by preventing the app process from completely shutting down after the main HWND is destroyed, leaving zombie processes that need to be manually killed via task manager. I've also seen a similar issue during app startup, but have found it more difficult to reproduce.

After some bisecting, it appears to have been introduced in this commit: https://github.com/Bill-Gray/PDCursesMod/commit/dd59e4cee70af08b86c111deba29cb7d2f795bb9 ... however, it's seemingly random and difficult to capture in a debugger.

I wish I had more information to share, but I don't. I speculate this it's due to the way the main critical section is managed; if you look at the code in the commit, sometimes the lock is acquired in one window message, then released in another. There's probably a condition where a lock is acquired, but the window is destroyed before the lock is released, then some time later another window message comes in and tries to acquire the lock.

In general it's probably dangerous to assume ordering of windows messages to acquire/release locks -- Windows has all kinds of weird legacy edge cases that may prevent some messages from being sent in some cases, on some platforms.

**Comment by Bill-Gray** [comment]
Hmmm... I see what you mean about the flow being difficult to follow.  In _most_ places,  you've got an `EnterCriticalSection()` followed by a few lines,  then `LeaveCriticalSection()`.  I didn't actually write these bits of the code,  so I may be off base here,  but the cases I don't immediately grok are :

-- Within `PDC_scr_open()`,  we `EnterCriticalSection()` and leave... when?

-- In the handling of WM_CLOSE,  if we're closing because of a shutdown-key event,  we call `final_cleanup` and `exit(0)` without ever leaving the critical section.  This is not normally going to happen unless you've set up a shutdown key (see comments in `pdcwin.h`) and I doubt it's what's happening here.  But it does look to me as if there ought to be a `LeaveCriticalSection()` call in there,  right after `final_cleanup()`.

-- In `PDC_scr_open()`,  there's a call to `EnterCriticalSection()` with no apparent corresponding `Leave`.

-- The entering and leaving of critical sections in `PDC_resize_screen()` might be perfectly fine,  but I could see this being a situation where mismatched messages might result in an entering without corresponding leaving situation.

About a decade ago,  I wrote WinGUI as a two-threaded system.  It always "sort of" worked,  but there were always odd edge cases that were difficult to reproduce or debug.  Then I realized I could make it single-threaded after all and,  with a large sigh of relief,  did so. (It was with another large sigh of relief that I adopted @wmcbrine's single-threaded X11 port.  I never really understood what was going on in the two-threaded version.)

Then in early 2020,  @loblolly986 added code returning us to a two-threaded WinGUI setup to get around problems with ugly redrawing and with temporary freeze-ups while resizing (issue #156).  I think I may take another look for solutions to those problems that don't involve multithreading... I might simply accept the ugliness,  and maybe temporary freeze-ups while resizing;  the freeze-ups while moving the window are hard to accept.  (For 'ugliness',  I think we could simply force the window contents not to be shown while resizing takes place.)
